feat(cloudformation): Add partial `NoEcho` parameter support

rominf opened the pull request:
<!-- Please refer to the contribution guidelines before raising a PR: https://github.com/localstack/localstack/blob/master/docs/CONTRIBUTING.md -->

<!-- Why am I raising this PR? Add context such as related issues, PRs, or documentation. -->
## Motivation

This commit introduces support for the `NoEcho` attribute in CloudFormation parameters. This change aligns LocalStack's behavior more closely with AWS CloudFormation, providing a more accurate local testing environment.

<!-- What changes does this PR make? How does LocalStack behave differently now? -->
## Changes
The implementation ensures that sensitive parameters marked with `NoEcho: true` are properly masked in responses from `describe-stacks` and `describe-change-set` operations, enhancing security and compliance.
    
- Masked `NoEcho` parameters in the Parameters section of `describe-stacks` output
- Applied similar masking logic to the Parameters section in `describe-change-set` output

Follow-up changes might include printing warnings when using `NoEcho` parameters, stating that this may expose sensitive information in the stack outputs and metadata. Users should ensure that sensitive information is not directly embedded in the template.

<!-- Optional section: How to test these changes? -->
<!--
## Testing

-->

<!-- Optional section: What's left to do before it can be merged? -->
<!--
## TODO

What's left to do:

- [ ] ...
- [ ] ...
-->

simonrw reviewed (commented):
I think this is a really good first attempt. Just a few comments in the code.

Did you consider the additional caveats when implementing this feature? 

![image](https://github.com/user-attachments/assets/f929ba63-3a43-44d4-9a09-5558582fd039)

[From the docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-requirements)

Would you tackle these here or in a follow up?

simonrw commented on localstack-core/localstack/aws/api/cloudformation/__init__.py:
```diff
@@ -1070,6 +1070,7 @@ class RollbackConfiguration(TypedDict, total=False):
 class Parameter(TypedDict, total=False):
     ParameterKey: Optional[ParameterKey]
     ParameterValue: Optional[ParameterValue]
+    NoEcho: Optional[NoEcho]
```
Sorry to be confusing about this, but this response is a mistake on my part. 

This file should not be updated, I was correct about that.

> Given these TypedDicts are marked total=False it does not mean the absence of this field is incorrect, just unfortunate when it comes to our typing information.

The rest of my comment regarding the specs actually is my misunderstanding. The `NoEcho` property is valid for templates, but this object is part of the API which does not support this parameter.

rominf commented:
@simonrw Thank you for your review and feedback.

I have addressed all code comments. As for handling the caveats related to `NoEcho` parameters, I want to ensure that my implementation aligns with AWS best practices. Specifically, here is what I have done:

1. **Outputs Section**: 
   - I have a test that ensures parameters referenced in the Outputs section are printed as-is in API responses.
   - This aligns with AWS documentation, which states that values in the Outputs section should not be masked.

2. **Metadata Section**:
   - Currently, my implementation does not mask any values in the Metadata section.
   - I plan to add tests to verify this behavior and ensure it remains consistent.

3. **Resource Metadata Attribute**:
   - Similarly, values in the Metadata attribute of resource definitions are not masked.
   - I plan to add tests to cover this aspect as well.

4. **Primary Identifiers**:
   - I understood that `NoEcho` parameters should not used in properties that form primary identifiers for resources.
   - If such a scenario is detected, I plan to print a warning message during stack creation or update operations.

5. **Warnings for Sensitive Data**:
   - While `NoEcho` does not mask data in the Metadata and Outputs sections and Resource Metadata Attribute, I will consider adding warnings when users create or update stacks with `NoEcho` parameters in these sections by modifying [resolve_dependencies](https://github.com/localstack/localstack/blob/3a32d7628ec1ce606abba8d79701c466fc6ad5e7/localstack-core/localstack/services/cloudformation/engine/template_utils.py#L26) function.
   - This aligns with AWS best practices to avoid embedding sensitive information directly in templates.

Would you like me to add specific tests for the Metadata section and Resource Metadata attribute? Additionally, should I proceed with printing warnings for `NoEcho` parameters used in places mentioned above?

simonrw commented:
@rominf I can see [one of our integration tests failing](https://app.circleci.com/pipelines/github/localstack/localstack/29805/workflows/4906d18d-19d9-49c0-af57-6692cc1922d7/jobs/263579/tests) with your latest commit. Could you fix this please?

simonrw reviewed (approved):
Thanks for resolving my suggestions, this PR is looking good 🎉 Thank you for your contribution!
